Sync and remove duplicate docs page (#2652)
fishtown-analytics/docs.getdbt.com: website/docs/docs/build/python-models.md

Instead of a final `select` statement, each Python model returns a final DataFrame. Each DataFrame operation is "lazily evaluated." In development, you can preview its data, using methods like `.show()` or `.head()`. When you run a Python model, the full result of the final DataFrame will be saved as a table in your data warehouse.

> dbt Python models have access to almost all of the same configuration options as SQL models. You can test them, document them, add `tags` and `meta` properties to them, grant access to their results to other users, and so on. You can select them by their name, their file path, their configurations, whether they are upstream or downstream of another model, or whether they have been modified compared to a previous project state.

Reviewer: ```suggestion
dbt Python models have access to almost all of the same configuration options as SQL models. You can test and document them, add `tags` and `meta` properties, and grant access to their results to other users. You can select them by their name, file path, configurations, whether they are upstream or downstream of another model, or if they have been modified compared to a previous project state.
```

Reviewer: I'm not sure "whether" is the correct choice for the last part. It's usually used when presenting a choice or selection of alternatives and "if" is used when establishing conditions.  Maybe we can take this one to the team.

def model(dbt, session):
    dbt.config(materialized = "incremental")
    df = dbt.ref("upstream_table")

    if dbt.is_incremental:

        # only new rows compared to max in current table
        max_from_this = f"select max(updated_at) from {dbt.this}"
        df = df.filter(df.updated_at >= session.sql(max_from_this).collect()[0][0])

        # or only rows from the past 3 days
        df = df.filter(df.updated_at >= F.dateadd("day", F.lit(-3), F.current_timestamp()))

    ...

    return df
```

</File>

</div>

<div warehouse="PySpark">

<File name='models/my_python_model.py'>

```python
import pyspark.sql.functions as F

def model(dbt, session):
    dbt.config(materialized = "incremental")
    df = dbt.ref("upstream_table")

    if dbt.is_incremental:

        # only new rows compared to max in current table
        max_from_this = f"select max(updated_at) from {dbt.this}"
        df = df.filter(df.updated_at >= session.sql(max_from_this).collect()[0][0])

        # or only rows from the past 3 days
        df = df.filter(df.updated_at >= F.date_add(F.current_timestamp(), F.lit(-3)))

    ...

    return df
```

</File>

</div>

</WHCode>

> Incremental models are supported on BigQuery with Dataproc for the `merge` incremental strategy. The `insert_overwrite` strategy is not yet supported.

Reviewer 2: Kind of feel like this should be up at line 223 in the form of a disclaimer of some sort. Not sure the best way to format that with our system but having this information located so far away from that statement could lead to confusion.

Author: thanks for flagging!

</File>

> Currently, Python functions defined in one dbt model can't be imported and reused in other models. Refer to [Code reuse](#code-reuse) for the potential patterns we're considering.

Reviewer: ```suggestion
Currently, Python functions defined in one dbt model can't be imported and reused in other models. Refer to [Code reuse](#code-reuse) for the potential patterns being considered.
```

### Using PyPI packages

> You can also define functions that depend on third-party packages so long as those packages are installed and available to the Python runtime on your data platform. See notes on "Installing Packages" for [specific data platforms](#specific-data-platforms).

Reviewer: Thank you for fixing these links and page title changes!

#### Code reuse

> Currently, you can't import or reuse Python functions defined in one dbt model in other models. This is something dbt Labs would like to support, so there are two patterns we're considering:

Reviewer 2: > defined in one dbt model in other models

The new syntax makes this really hard for me to parse. I actual prefer the original way this sentence was worded. The rest of the change is lovely.

As a general rule, if there's a transformation you could write equally well in SQL or Python, we believe that well-written SQL is preferable: it's more accessible to a greater number of colleagues, and it's easier to write code that's performant at scale. If there's a transformation you _can't_ write in SQL, or where ten lines of elegant and well-annotated Python could save you 1000 lines of hard-to-read Jinja-SQL, Python is the way to go.

> ## Specific data platform {#specific-data-platforms}

Author: ```suggestion
## Specific data platforms {#specific-data-platforms}
```
